### PR TITLE
Complete v13 type definitions, support metadata decoding and serde

### PR DESCRIPTION
**Pull request by ascjones**
More features for substrate compatibility and for use of the resulting encoded metadata in https://github.com/ascjones/chameleon.

- Support SCALE decoding of v13 metadata
- Support serde serialization of v13 metadata
- Add definintion for extrinsics and their signed extensions

### CODE REVIEW

File: frame-metadata/src/v13.rs [hunk 1]
```diff
@@ -74,36 +79,75 @@ impl RuntimeMetadataV13 {
...
 pub struct ModuleMetadata<T: Form = MetaForm> {
 	pub name: T::String,
 	pub storage: Option<Vec<StorageMetadata<T>>>,
 	pub calls: Option<Vec<FunctionMetadata<T>>>,
 	pub event: Option<Vec<EventMetadata<T>>>,
 	// pub constants: DFnA<ModuleConstantMetadata>,
 	pub errors: Vec<ErrorMetadata<T>>,
+	// /// Define the index of the module, this index will be used for the encoding of module event,
+	// /// call and origin variants.
+	// pub index: u8,
```
Comment (dvdplm) [inline frame-metadata/src/v13.rs:150]: Also outstanding?

**Review by dvdplm — COMMENTED** [review]
There are two members of `ModuleMetadata` left out. I imagine that's on purpose but I'm not sure why – can you elaborate?

File: frame-metadata/src/v13.rs [hunk 2]
```diff
@@ -74,36 +79,75 @@ impl RuntimeMetadataV13 {
...
+#[cfg_attr(
+	feature = "std",
+	serde(bound(serialize = "T::Type: Serialize, T::String: Serialize"))
+)]
 pub struct ModuleMetadata<T: Form = MetaForm> {
 	pub name: T::String,
 	pub storage: Option<Vec<StorageMetadata<T>>>,
 	pub calls: Option<Vec<FunctionMetadata<T>>>,
 	pub event: Option<Vec<EventMetadata<T>>>,
 	// pub constants: DFnA<ModuleConstantMetadata>,
```
[the remark this answers is not in the document]
Comment (ascjones) [reply frame-metadata/src/v13.rs:146]: Yeah it's outstanding, I have been going through the fields one by one and adding them in on the substrate side. That one is next on the list.

File: frame-metadata/src/v13.rs (hunk 1 above)
Comment (ascjones) [reply frame-metadata/src/v13.rs:150]: As above

**Comment by ascjones** [comment]
> There are two members of ModuleMetadata left out. I imagine that's on purpose but I'm not sure why – can you elaborate?

Indeed those still need to be done, I was going to wait until I'd added those before submitting this PR, but decided to put it up with the changes so far once I knew the decoding fixes might be required in the meantime.

I will flesh out these fields/types for now for completeness on this PR, and then just populate them with default values on the substrate side.

File: frame-metadata/src/v12.rs [hunk 3]
```diff
@@ -362,11 +368,14 @@ pub struct ExtrinsicMetadata {
...
 #[cfg_attr(feature = "std", derive(Decode, Serialize, Debug))]
-pub struct RuntimeMetadataV12 {
+#[cfg_attr(feature = "std", serde(bound(serialize = "S: Serialize")))]
+pub struct RuntimeMetadataV12<S = ()> {
 	/// Metadata of all the modules.
 	pub modules: DecodeDifferentArray<ModuleMetadata>,
 	/// Metadata of the extrinsic.
 	pub extrinsic: ExtrinsicMetadata,
+	/// Marker for dummy type parameter required by v13 but not used here.
+	pub marker: PhantomData<S>,
```
[the remark this answers is not in the document]
Comment (ascjones) [reply frame-metadata/src/v12.rs:378]: There's no constructor so it needs to be

**Comment by ascjones** [comment]
> I will flesh out these fields/types for now for completeness on this PR, and then just populate them with default values on the substrate side.

Done

**Review by dvdplm — APPROVED** [review]

**Pull request merged**